gcc/rust/rust-diagnostics.h

```diff
@@ -51,15 +51,1040 @@
+  E0615,
+  E0616,
+  E0617,
+  E0618,
+  E0619, // this error code is no longer emitted by the compiler
+  E0620,
+  E0621,
+  E0622,
+  E0623,
+  E0624,
+  E0625,
+  E0626,
+  E0627,
+  E0628,
+  E0631,
+  E0632, // this error code is no longer emitted by the compiler
+  E0633, // this error code is no longer emitted by the compiler
+  E0634,
+  E0635,
+  E0636,
+  E0637,
+  E0638,
+  E0639,
+  E0640,
+  E0641,
+  E0642,
+  E0643,
+  E0644,
+  E0646,
+  E0647,
+  E0648,
+  E0657,
+  E0658,
+  E0659,
+  E0660, // this error code is no longer emitted by the compiler
+  E0661, // this error code is no longer emitted by the compiler
+  E0662, // this error code is no longer emitted by the compiler
+  E0663, // this error code is no longer emitted by the compiler
+  E0664, // this error code is no longer emitted by the compiler
+  E0665, // this error code is no longer emitted by the compiler
+  E0666,
+  E0667,
+  E0668, // this error code is no longer emitted by the compiler
+  E0669, // this error code is no longer emitted by the compiler
+  E0670,
+  E0671, // this error code is no longer emitted by the compiler
+  E0687, // this error code is no longer emitted by the compiler
+  E0688, // this error code is no longer emitted by the compiler
+  E0689,
+  E0690,
+  E0691,
+  E0692,
+  E0693,
+  E0695,
+  E0696,
+  E0697,
+  E0698,
+  E0699,
+  E0700,
+  E0701,
+  E0703,
+  E0704,
+  E0705,
+  E0706,
+  E0708,
+  E0710,
+  E0711,
+  E0712,
+  E0713,
+  E0714,
+  E0715,
+  E0716,
+  E0717,
+  E0718,
+  E0719,
+  E0720,
+  E0722,
+  E0724,
+  E0725,
+  E0726,
+  E0727,
+  E0728,
+  E0729,
+  E0730,
+  E0731,
+  E0732,
+  E0733,
+  E0734,
+  E0735,
+  E0736,
+  E0737,
+  E0739,
+  E0740,
+  E0741,
+  E0742,
+  E0743,
+  E0744,
+  E0745,
+  E0746,
+  E0747,
+  E0748,
+  E0749,
+  E0750,
+  E0751,
+  E0752,
+  E0753,
+  E0754,
+  E0755,
+  E0756,
+  E0757,
+  E0758,
+  E0759, // this error code is no longer emitted by the compiler
+  E0760, // this error code is no longer emitted by the compiler
+  E0761,
+  E0762,
+  E0763,
+  E0764,
+  E0765,
+  E0766,
+  E0767,
+  E0768,
+  E0769,
+  E0770,
+  E0771,
+  E0772, // this error code is no longer emitted by the compiler
+  E0773,
+  E0774,
+  E0775,
+  E0776,
+  E0777,
+  E0778,
+  E0779,
+  E0780,
+  E0781,
+  E0782,
+  E0783,
+  E0784,
+  E0785,
+  E0786,
+  E0787,
+  E0788,
+  E0789,
+  E0790,
+  E0791,
+  E0792,
+  E0793,
+  E0794
+};
+
+// Custom hash function for ErrorCode, for older version of gcc
+namespace std {
+    template <>
+    struct hash<ErrorCode> {
+        size_t operator()(const ErrorCode& custom_hash) const {
+            return static_cast<size_t>(custom_hash);
+        }
+    };
+}
+
+// see https://github.com/Rust-GCC/gccrs/pull/2468
+#define XSTR(A) STR(A)
+#define STR(A) #A
+#define ERROR_CODE(NUM) E##NUM
+#define TABLE_TO_MAP(NUM) {ErrorCode::ERROR_CODE(NUM), XSTR(ERROR_CODE(NUM))} 
 
-  const char *m_str;
+static std::unordered_map<ErrorCode,const char *> ErrorCodeString = {
```

Comment:
these macros are very good. I agree with @P-E-P that I think it is the way to go. We could clean it up even further by having a `error_codes.incl` file or something, which contains the content of the `ErrorCode` enum wrapped in an x-macro. So something like:

```
  ERROR(E0023),
  ERROR(E0025),
  ERROR(E0026),
  ```
  
  and then have the enum be defined as
  
  ```c++
  // do nothing
  #define ERROR 
  enum ErrorCode {
  #include "error_codes.incl"
  };
  #undef ERROR
  ```
 and the table
 ```c++
 #define ERROR(e) {ErrorCode::##E, ##E}
 static std::unordered_map<ErrorCode,const char *> ErrorCodeString = {
 #include "error_codes.incl"
 };
 #undef ERROR
```

for example. this way, we only have to edit one file when a new error code gets added. we can also use this file later for things like error messages, translations, etc.

That can all be done in a later PR tho